compendium/LogServiceImpl/src/LogServiceImpl.cpp

```diff
@@ -6,110 +6,198 @@
 #include "cppmicroservices/ServiceReference.h"
 
 #include "LogServiceImpl.hpp"
+#include "LoggerFactoryImpl.hpp"
+#include "LoggerImpl.hpp"
 
-namespace cppmicroservices
-{
-    namespace logservice
-    {
-        std::string
-        GetExceptionMessage(std::exception_ptr const& ex)
+namespace cppmicroservices::logservice
+{     
+        LogServiceImpl::LogServiceImpl(std::string const& loggerName)
         {
-            std::string message = "\nException logged: ";
-            if (ex)
-            {
-                std::ostringstream stream;
-                try
-                {
-                    std::rethrow_exception(ex);
-                }
-                catch (std::exception const& e)
-                {
-                    message += std::string(typeid(e).name()) + " : " + e.what();
-                }
-            }
-            else
-            {
-                message += "none";
-            }
-
-            return message;
-        }
-
-        std::string
-        GetServiceReferenceInfo(ServiceReferenceBase const& sr)
+            logger = getLogger(loggerName);    
+	}
+   
+        void
+        LogServiceImpl::Log(SeverityLevel level, std::string const& message)
         {
-            std::ostringstream stream;
-            stream << "\nServiceReference: " << sr;
-            return stream.str();
-        }
 
-        LogServiceImpl::LogServiceImpl(std::string const& loggerName)
-        {
-            auto sink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
-            m_Logger = std::make_shared<spdlog::logger>(loggerName, std::move(sink));
-            m_Logger->set_pattern("[%T] [%P:%t] %n (%^%l%$): %v");
-            m_Logger->set_level(spdlog::level::trace);
-        }
+           auto currLogger = std::atomic_load(&logger);
+           if (!currLogger)
+           {
+               return;
+           }
+
+           switch (level)
+           {
+               case SeverityLevel::LOG_DEBUG:
+               {
+		   currLogger->debug(message);
```

Comment:
I think having this switch statement 3 times in 3 functions is perhaps not ideal. Rather what if we just had a 'impl' of the `Log` call that looked like this:

```
template<typename... Args>
void LogServiceImpl::logImpl(SeverityLevel level, Args&&... args) {
    auto currLogger = std::atomic_load(&logger);
    if (!currLogger) {
        return;
    }

    switch (level) {
        case SeverityLevel::LOG_DEBUG:
            currLogger->debug(std::forward<Args>(args)...);
            break;
        case SeverityLevel::LOG_INFO:
            currLogger->info(std::forward<Args>(args)...);
            break;
        case SeverityLevel::LOG_WARNING:
            currLogger->warn(std::forward<Args>(args)...);
            break;
        case SeverityLevel::LOG_ERROR:
            currLogger->error(std::forward<Args>(args)...);
            break;
        default:
            currLogger->trace(std::forward<Args>(args)...);
            break;
    }
}
```

**Note**: This would need to be in hpp as it is templated

And then each of the other log calls could simply forward their arguments:

```
void LogServiceImpl::Log(SeverityLevel level, std::string const& message) {
    logImpl(level, message);
}

void LogServiceImpl::Log(SeverityLevel level, std::string const& message, const std::exception_ptr ex) {
    logImpl(level, message, ex);
}

void LogServiceImpl::Log(ServiceReferenceBase const& sr, SeverityLevel level, std::string const& message) {
	    logImpl(level, message, sr);
}
```

Reply:
Thanks for the feedback.
However, this change would necessitate making all the log methods in the Logger class (e.g., debug, error, info, etc.) template functions that can accept variadic arguments. This would mean that these templates, along with their implementations, would reside in the service interface, resulting in a loss of the loose coupling between service consumers and the service implementation.

Reply:
Hi Archita, thanks for the response. 

I am not sure this would require that all the log methods take variadic arguments. 

I cloned your branch and imjplemented my suggestion such that the LogServiceImpl class made all its calls like this:

```
    void
    LogServiceImpl::Log(SeverityLevel level, std::string const& message, std::exception_ptr const ex)
    {
        logImpl(level, message, ex);
    }
```

and in LogServiceImpl.hpp:

```
            template <typename... Args>
            void
            logImpl(SeverityLevel level, Args&&... args)
            {
                auto currLogger = std::atomic_load(&logger);
                if (!currLogger)
                {
                    return;
                }

                switch (level)
                {
                    case SeverityLevel::LOG_DEBUG:
                        currLogger->debug(std::forward<Args>(args)...);
                        break;
                    case SeverityLevel::LOG_INFO:
                        currLogger->info(std::forward<Args>(args)...);
                        break;
                    case SeverityLevel::LOG_WARNING:
                        currLogger->warn(std::forward<Args>(args)...);
                        break;
                    case SeverityLevel::LOG_ERROR:
                        currLogger->error(std::forward<Args>(args)...);
                        break;
                    default:
                        currLogger->trace(std::forward<Args>(args)...);
                        break;
                }
            }
```
And this compiled and passed all tests without any changes to the Logger class functions. 

Let me know if I am missing something here, but I think this should work.